consensus/state_processing/src/common/update_progressive_balances_cache.rs

```diff
@@ -0,0 +1,152 @@
+    ParticipationFlags, ProgressiveBalancesCache,
+};
+
+/// Initializes the `ProgressiveBalancesCache` cache using balance values from the
+/// `ParticipationCache`. If the optional `&ParticipationCache` is not supplied, it will be computed
+/// from the `BeaconState`.
+pub fn initialize_progressive_balances_cache<E: EthSpec>(
+    state: &mut BeaconState<E>,
+    maybe_participation_cache: Option<&ParticipationCache>,
+    spec: &ChainSpec,
+) -> Result<(), BeaconStateError> {
+    if !is_progressive_balances_enabled(state)
+        || state.progressive_balances_cache().is_initialized()
+    {
+        return Ok(());
+    }
+
+    let participation_cache = match maybe_participation_cache {
+        Some(cache) => Cow::Borrowed(cache),
+        None => Cow::Owned(ParticipationCache::new(state, spec)?),
+    };
+
+    let previous_epoch_target_attesting_balance = participation_cache
+        .previous_epoch_target_attesting_balance_raw()
+        .map_err(|e| BeaconStateError::ParticipationCacheError(format!("{:?}", e)))?;
+
+    let current_epoch_target_attesting_balance = participation_cache
+        .current_epoch_target_attesting_balance_raw()
+        .map_err(|e| BeaconStateError::ParticipationCacheError(format!("{:?}", e)))?;
+
+    let current_epoch = state.current_epoch();
+    state.progressive_balances_cache_mut().initialize(
+        current_epoch,
+        previous_epoch_target_attesting_balance,
+        current_epoch_target_attesting_balance,
+    );
+
+    update_progressive_balances_metrics(state.progressive_balances_cache())?;
+
+    Ok(())
+}
+
+/// Updates the `ProgressiveBalancesCache` when a new target attestation has been processed.
+pub fn update_progressive_balances_on_attestation<T: EthSpec>(
+    state: &mut BeaconState<T>,
+    epoch: Epoch,
+    validator_index: usize,
+) -> Result<(), BlockProcessingError> {
+    if is_progressive_balances_enabled(state) {
+        let validator = state.get_validator(validator_index)?;
+        if !validator.slashed {
+            let validator_effective_balance = validator.effective_balance;
+            state
+                .progressive_balances_cache_mut()
+                .on_new_target_attestation(epoch, validator_effective_balance)?;
+
+            update_progressive_balances_metrics(state.progressive_balances_cache())?;
+        }
+    }
+    Ok(())
+}
+
+/// Updates the `ProgressiveBalancesCache` when a target attester has been slashed.
+pub fn update_progressive_balances_on_slashing<T: EthSpec>(
+    state: &mut BeaconState<T>,
+    validator_index: usize,
+) -> Result<(), BlockProcessingError> {
+    if is_progressive_balances_enabled(state) {
+        let is_previous_epoch_target_attester = match state.previous_epoch_participation() {
+            Ok(previous_epoch_participation) => {
+                is_target_attester_in_epoch::<T>(previous_epoch_participation, validator_index)?
+            }
+            Err(_) => false,
```

Comment:
I think we can safely `?` the `state.previous_epoch_participation` rather than handling the error, because the `is_progressive_balances_enabled` function should guard against running this code for the phase0/base fork.

Reply:
good point - I'll update this.